File: backend-node/src/group/group.controller.ts
import { Controller, Get, Post, Body, Param, Session, UseGuards, HttpCode, HttpStatus, BadRequestException, NotFoundException } from '@nestjs/common';
import { GroupService } from './group.service';
import { SessionGuard } from '../auth/session.guard';
import { PermissionsGuard } from '../auth/permissions.guard';
import { Permissions } from '../auth/permissions.decorator';
import { PERMISSIONS } from '../auth/permission.constants';

@Controller('group')
@UseGuards(SessionGuard)
export class GroupController {
  constructor(private readonly groupService: GroupService) {}

  @Post()
  async createGroup(
    @Body() body: { name: string; description?: string },
    @Session() session: Record<string, any>,
  ) {
    const group = await this.groupService.createGroup(
      body.name,
      body.description || '',
      session.userId,
    );

    return {
      success: true,
      message: '群组创建成功',
      group,
    };
  }

  @Get()
  async getUserGroups(@Session() session: Record<string, any>) {
    const groups = await this.groupService.getUserGroups(session.userId);
    return {
      success: true,
      groups,
    };
  }

  @Get(':groupId')
  async getGroupDetails(
    @Param('groupId') groupId: string,
    @Session() session: Record<string, any>,
  ) {
    const group = await this.groupService.getGroupDetails(groupId, session.userId);
    return {
      success: true,
      group,
    };
  }

  @Post(':groupId/invite')
  @HttpCode(HttpStatus.OK)
  async inviteUser(
    @Param('groupId') groupId: string,
    @Body() body: { userId: string },
    @Session() session: Record<string, any>,
  ) {
    await this.groupService.inviteUser(groupId, body.userId, session.userId);
    return {
      success: true,
      message: '邀请成功',
    };
  }

  @Post(':groupId/members/:userId/delete')
  async removeMember(
    @Param('groupId') groupId: string,
    @Param('userId') userId: string,
    @Session() session: Record<string, any>,
  ) {
    await this.groupService.removeMember(groupId, userId, session.userId);
    return {
      success: true,
      message: '成员已移除',
    };
  }

  @Post(':groupId')
  async updateGroup(
    @Param('groupId') groupId: string,
    @Body() body: { name: string; description?: string },
    @Session() session: Record<string, any>,
  ) {
    const group = await this.groupService.updateGroup(
      groupId,
      body.name,
      body.description || '',
      session.userId,
    );

    return {
      success: true,
      message: '群组信息已更新',
      group,
    };
  }

  @Post(':groupId/delete')
  async deleteGroup(
    @Param('groupId') groupId: string,
    @Session() session: Record<string, any>,
  ) {
    await this.groupService.deleteGroup(groupId, session.userId);
    return {
      success: true,
      message: '群组已删除',
    };
  }

  @Post(':groupId/members/:userId/permissions')
  @UseGuards(PermissionsGuard)
  @Permissions(PERMISSIONS.GROUP_MANAGE)
  async updateMemberPermissions(
    @Param('groupId') groupId: string,
    @Param('userId') userId: string,
    @Body() body: { permissions: string[] },
  ) {
    if (!Array.isArray(body.permissions)) {
      throw new BadRequestException('permissions 必须是数组');
    }
    await this.groupService.updateMemberPermissions(groupId, userId, body.permissions);
    return {
      success: true,
      message: '权限更新成功',
    };
  }

  @Get(':groupId/members/:userId/permissions')
  @UseGuards(PermissionsGuard)
  @Permissions(PERMISSIONS.GROUP_MANAGE)
  async getMemberPermissions(
    @Param('groupId') groupId: string,
    @Param('userId') userId: string,
  ) {
    const permissions = await this.groupService.getMemberPermissions(groupId, userId);
    return {
      success: true,
      permissions,
    };
  }
}

File: backend-java/mvgo-api/src/main/java/com/mvgo/api/controller/GroupController.java
package com.mvgo.api.controller;

import java.util.List;
import java.util.Map;

import com.mvgo.business.service.GroupService;
import com.mvgo.common.Result;

import org.springframework.web.bind.annotation.*;

/**
 * 群组管理接口（迁自 Node GroupController）.
 *
 * @since 1.0.0
 */
@RestController
@RequestMapping("/api/group")
public class GroupController {

    private final GroupService groupService;

    public GroupController(GroupService groupService) {
        this.groupService = groupService;
    }

    /** 创建群组 */
    @PostMapping
    public Result<Map<String, Object>> createGroup(@RequestBody Map<String, Object> body) {
        String name = (String) body.get("name");
        String description = (String) body.get("description");
        String adminId = (String) body.get("adminId");
        return Result.success(groupService.createGroup(name, description, adminId));
    }

    /** 获取用户所属群组列表 */
    @GetMapping
    public Result<List<Map<String, Object>>> getUserGroups(@RequestParam("userId") String userId) {
        return Result.success(groupService.getUserGroups(userId));
    }

    /** 获取群组详情 */
    @GetMapping("{groupId}")
    public Result<Map<String, Object>> getGroupDetails(
            @PathVariable String groupId,
            @RequestParam("userId") String userId) {
        return Result.success(groupService.getGroupDetails(groupId, userId));
    }

    /** 邀请用户 */
    @PostMapping("{groupId}/invite")
    public Result<Map<String, Object>> inviteUser(
            @PathVariable String groupId,
            @RequestBody Map<String, Object> body) {
        String email = (String) body.get("email");
        String inviterId = (String) body.get("inviterId");
        return Result.success(groupService.inviteUser(groupId, email, inviterId));
    }

    /** 移除成员 */
    @PostMapping("{groupId}/members/{userId}/delete")
    public Result<Void> removeMember(
            @PathVariable String groupId,
            @PathVariable("userId") String memberUserId,
            @RequestParam("operatorId") String operatorId) {
        groupService.removeMember(groupId, memberUserId, operatorId);
        return Result.success(null);
    }

    /** 更新群组 */
    @PostMapping("{groupId}")
    public Result<Map<String, Object>> updateGroup(
            @PathVariable String groupId,
            @RequestBody Map<String, Object> updates,
            @RequestParam("userId") String userId) {
        var group = groupService.updateGroup(groupId, updates, userId);
        return Result.success(Map.of("_id", group.getId(), "name", group.getName(), "updatedAt", group.getUpdatedAt()));
    }

    /** 删除群组 */
    @PostMapping("{groupId}/delete")
    public Result<Void> deleteGroup(
            @PathVariable String groupId,
            @RequestParam("userId") String userId) {
        groupService.deleteGroup(groupId, userId);
        return Result.success(null);
    }

    /** 更新成员权限 */
    @PostMapping("{groupId}/members/{userId}/permissions")
    public Result<Map<String, Object>> updateMemberPermissions(
            @PathVariable String groupId,
            @PathVariable("userId") String memberUserId,
            @RequestBody Map<String, Object> body) {
        String operatorId = (String) body.get("operatorId");
        @SuppressWarnings("unchecked")
        List<String> permissions = (List<String>) body.get("permissions");
        return Result.success(groupService.updateMemberPermissions(groupId, memberUserId, permissions, operatorId));
    }

    /** 获取成员权限 */
    @GetMapping("{groupId}/members/{userId}/permissions")
    public Result<Map<String, Object>> getMemberPermissions(
            @PathVariable String groupId,
            @PathVariable("userId") String memberUserId) {
        return Result.success(groupService.getMemberPermissions(groupId, memberUserId));
    }
}